File: Ada/livro-ds/mergepv.c
/* (C) 1999, Arthur Vargas Lopes */
/* Arquivo fonte MergePV.c       */

#include "MergePV.h"

/* Área de controle de vias */
Via DV[P];

void Swap(Rec *X, Rec *Y)
{
	Rec Temp;
	
	Temp = *X;
	*X   = *Y;
	*Y   = Temp;

} /* Swap */

/* Algoritmo 8.5 */
void Shell_Sort(Rec *V,int n)
{
	int D;
	int Ref;
	int Base;

	D = n / 2;
	while(D > 0)  {
			for(Ref = D; Ref < n; ++Ref) {
				Base = Ref - D;
				while (Base >= 0) 
					if (strcmp(V[Base].Chave,V[Base+D].Chave) > 0)  {
						Swap(&V[Base],&V[Base+D]);
						Base = Base - D;
						}
					else
						break;
				}
			D = D / 2;
			}

} /* Sort_Shell */

void Distribui_F(FILE *A_In, int Pi)
{
	Rec  V[M];
	Data R;
	int  I;
	static Pos = 0;

	/* Monta o nome do arquivo de saida */
	strcpy(DV[Pi].Fn,"F .DAT");
	DV[Pi].Fn[1] = Pi + '0';

	/* Inicializa indicador de fim de arquivo */
	DV[Pi].Fim   = 0;

	/* A última via pode ter menos
	  que M registros */
	if (P == (Pi + 1)) 
		DV[Pi].n = N - ((P - 1) * M);
	else
		DV[Pi].n = M;

	/* Armazena as primeiras M chaves e
	  seus respectivos tags em V */
	for(I = 0; I < DV[Pi].n; ++I) {
		if (fread(&R,sizeof(Data),1,A_In) != 1) {
			printf("Nao pode ler o registro %d do arquivo '%s'\n"
					, I+1, Entrada);
			exit(1);
			}
		strcpy(V[I].Chave,R.Nome);
		V[I].Tag = Pos;
		Pos      = Pos + 1;
		}

	/* Ordena as chaves de V */
	Shell_Sort(V,DV[Pi].n);

	/* Grava o conteúdo de V no arquivo Fp */
	DV[Pi].F = fopen(DV[Pi].Fn,"wb");
	if (DV[Pi].F == NULL)  {
		printf("Nao pode criar o arquivo '%s'\n", DV[Pi].Fn);
		exit(1);
		}

	for(I = 0; I < DV[Pi].n; ++I) 
		if (fwrite(&V[I],sizeof(Rec),1,DV[Pi].F) != 1) {
			printf("Nao pode gravar o registro %d do arquivo '%s'\n"
					, I+1, DV[Pi].Fn);
			exit(1);
			}

	fclose(DV[Pi].F);
	
} /* Distribui_F */



void Fase(char A2[])
{
	int  I;
	int  PVia;
	int  MenorVia;
	char MenorKey[NOME_SIZE+1];
	int  V_Ativas;
	FILE *A_Out;

	/* Reabre todas as vias */
	for(I = 0; I < P; ++I) {
		DV[I].F = fopen(DV[I].Fn,"rb");
		if (DV[I].F == NULL)  {
			printf("Nao pode encontar o arquivo '%s'\n", DV[I].Fn);
			exit(1);
			}
		}

	/* Abre o arquivo de saída */
	A_Out = fopen(A2,"wb");
	if (A_Out == NULL)  {
		printf("Nao pode criar o arquivo '%s'\n",A2);
		exit(1);
		}

	/* Le um registro de cada via */
	for(I = 0; I < P; ++I)
		if (fread(&DV[I].R,sizeof(Rec),1,DV[I].F) != 1) {
			printf("Nao pode ler o registro 1 do arquivo '%s'\n"
				, DV[I].Fn);
			exit(1);
			}

	V_Ativas = P;

	/* Grave o tag do registro que contiver
      a menor chave e leia um novo registro
	  do arquivo associado com esta chave
	  enquanto existirem registros a serem
	  lidos em pelo menos duas vias ativas */

	while (V_Ativas > 1) {

		/* Localiza a 1a via ativa */
		PVia = 0;
		while ((PVia < P) && (DV[PVia].Fim)) 
			PVia = PVia + 1;

		/* Seleciona a menor chave */
		MenorVia = PVia;
		strcpy(MenorKey,DV[PVia].R.Chave);
		PVia = PVia + 1;
		while (PVia < P)  {
			if (!DV[PVia].Fim)
				if (strcmp(DV[PVia].R.Chave,MenorKey) < 0) {
					MenorVia = PVia;
					strcpy(MenorKey,DV[PVia].R.Chave);
					}
			PVia = PVia + 1;
			}

		/* Grava o tag da via com a menor chave
		  no arquivo de referência */
		if (fwrite(&DV[MenorVia].R.Tag,sizeof(int),1,A_Out) != 1) {
			printf("Nao pode gravar registro do arquivo '%s'\n"
					, A2);
			exit(1);
			}

		DV[MenorVia].n = DV[MenorVia].n - 1;

		if (DV[MenorVia].n == 0)  {
			DV[MenorVia].Fim = 1;
			V_Ativas         = V_Ativas - 1;
			}
		else /* Le mais um registro desta via */
			if (fread(&DV[MenorVia].R,sizeof(Rec),1,DV[MenorVia].F) != 1) {
				printf("Nao pode ler o registro 1 do arquivo '%s'\n"
					, DV[MenorVia].Fn);
				exit(1);
				}
		}

	/* Neste ponto apenas uma das vias
	  contém chaves cujos tags ainda não foram
	  transferidos para o arquivo de referência */

	/* Localiza a via com tags pendentes */
	PVia = 0;
	while ((PVia < P) && (DV[PVia].Fim))  
		PVia = PVia + 1;

	while (DV[PVia].n > 0)  {

		if (fwrite(&DV[PVia].R.Tag,sizeof(int),1,A_Out) != 1) {
			printf("Nao pode gravar registro do arquivo '%s'\n"
					, A2);
			exit(1);
			}

		DV[PVia].n = DV[PVia].n - 1;

		if (DV[PVia].n > 0)
			/* Le mais um registro desta via */
			if (fread(&DV[PVia].R,sizeof(Rec),1,DV[PVia].F) != 1) {
				printf("Nao pode ler o registro 1 do arquivo '%s'\n"
					, DV[PVia].Fn);
				exit(1);
				}
		}

	/* Fecha e remove todas as vias */
	for(I = 0; I < P; ++I)  {
		fclose(DV[I].F);
		unlink(DV[I].Fn);
		}

	/* Fecha o arquivo de saída */
	fclose(A_Out);

} /* Fase */


/* Algoritmo 9.2 */
void Merge_P_Vias(char A[], char A2[])
{
	FILE *A_In;
	int I;

	/* Abre o arquivo de entrada */
	A_In = fopen(A,"rb");
	if (A_In == NULL)  {
		printf("Nao pode encontar o arquivo '%s'\n", A);
		exit(1);
		}

	/* Ordena e distribui N chaves em P vias */
	for(I = 0; I < P; ++I)
		Distribui_F(A_In,I);

	/* Realiza a operação de merge com F1 e F2
	   gerando o arquivo de referência         */
	Fase(A2);

} /* Merge_P_Vias */

File: Ada/livro-ds/mergepv.pas
{ (C) 1999, Arthur Vargas Lopes }
{ Arquivo fonte MergePV.Pas     }

PROGRAM MergePV;

	USES Crt;

CONST
	N       = 26 * 26;     { Número de registros a serem
				 ordenados }

	Pv	= 4;           { Número de vias -- sub-arquivos }

	M       = N DIV Pv;    { Comprimento do intervalo da
				 área de ordenação }

	Entrada = 'c:A.DAT';    { Nome do arquivo de entrada }
	Saida   = 'd:A2.DAT';   { Nome do arquivo de saída }

TYPE
	Key   = ARRAY[1..2] OF CHAR; { Chave de ordenação }

	Data  = RECORD        { Registro do arquivo de dados }
		Nome    : Key;
		Idade   : INTEGER;
		Salario : REAL;
		END;

	Rec   = RECORD        { Registro da area de ordenação }
		Chave : Key;
		Tag   : INTEGER;
		END;

	Arq   = FILE OF Data; { Destritor para arquivo a ser ordenado }
	Arq_T = FILE OF Rec;  { Destritor para de referência ordenado }

	{ Descritor de cada Via }
	Via   = RECORD
		R   : Rec;        { Buffer para merge }
		F   : Arq_T;      { Descritor do arquivo }
		N   : INTEGER;    { Número de registros da via}
		Fn  : STRING[12]; { Nome externo do arquivo }
		Fim : BOOLEAN;    { TRUE se encontrou EOF }
		END;

	Vetor = ARRAY[1..M] OF Rec;  { Área de ordenação }

	Vias  = ARRAY[1..Pv] OF Via; { Área de controle de vias }

VAR
	I, J, K   : INTEGER;

	Registro  : Data;
	Registro2 : Data;

	A_In      : Arq;             { Descritor do arquivo de entrada }
	A_Out     : FILE OF INTEGER; { Descritor do arquivo de saída }

	V         : Vetor;           { Armazena área de ordenação }

	DV        : Vias;            { Área descritora de vias }

	PROCEDURE Swap(VAR X, Y : Rec);
	VAR
		Temp : Rec;
	BEGIN
		Temp := X;
		X    := Y;
		Y    := Temp;
	END; { Swap }

	{ Algoritmo 8.5 }
	PROCEDURE Shell_Sort(N : INTEGER);
	VAR
		D    : INTEGER;
		Ref  : INTEGER;
		Base : INTEGER;
	BEGIN
		D := N DIV 2;

		WHILE D > 0 DO BEGIN
			FOR Ref := D + 1 TO N DO BEGIN
				Base := Ref - D;
				WHILE Base > 0 DO
					IF V[Base].Chave > V[Base+D].Chave THEN BEGIN
						Swap(V[Base],V[Base+D]);
						Base := Base - D;
						END
					ELSE
						Base := 0;
				END;
			D := D DIV 2;
			END;

	END; { Shell_Sort }

	{ Algoritmo 9.2 }
	PROCEDURE Merge_P_Vias(A : STRING;A2 : STRING);
	VAR
		I, J : INTEGER;
		R    : Data;
		Pos  : INTEGER;

		PROCEDURE Distribui_F(P : INTEGER);
		VAR
			I : INTEGER;
		BEGIN
			{ Monta o nome do arquivo de saida }
			DV[P].FN    := 'F .DAT';
			DV[P].FN[2] := Chr(P + Ord('0'));

			{ Inicializa indicador de fim de arquivo }
			DV[P].Fim   := FALSE;

			{ A última via pode ter menos
			  que M registros }
			IF P = Pv THEN
				DV[P].N := N - ((P - 1) * M)
			ELSE
				DV[P].N := M;

			{ Armazena as primeiras M chaves e
			  seus respectivos tags em V }
			FOR I := 1 TO DV[P].N DO BEGIN
				Read(A_In, R);
				V[I].Chave := R.Nome;
				Pos        := Pos + 1;
				V[I].Tag   := Pos;
				END;

			{ Ordena as chaves de V }
			Shell_Sort(DV[P].N);

			{ Grava o conteúdo de V no arquivo Fp }
			Assign(DV[P].F, DV[P].Fn);
			ReWrite(DV[P].F);

			FOR I := 1 TO DV[P].N DO
				Write(DV[P].F, V[I]);

			Close(DV[P].F);

		END; { Distribui_F }

		PROCEDURE Fase;
		VAR
			I        : INTEGER;
			PVia     : INTEGER;
			MenorVia : INTEGER;
			MenorKey : Key;
			V_Ativas : INTEGER;
		BEGIN
			{ Reabre todas as vias }
			FOR I := 1 TO Pv DO
				Reset(DV[I].F);

			{ Abre o arquivo de saída }
			Assign(A_Out,A2);
			ReWrite(A_Out);


			{ Le um registro de cada via }
			FOR I := 1 TO Pv DO
				Read(DV[I].F, DV[I].R);

			V_Ativas := Pv;

			{ Grave o tag do registro que contiver
			  a menor chave e leia um novo registro
			  do arquivo associado com esta chave
			  enquanto existirem registros a serem
			  lidos em pelo menos duas vias ativas }

			WHILE(V_Ativas > 1) DO BEGIN

				{ Localiza a 1a via ativa }
				PVia := 1;
				WHILE (PVia <= Pv) AND (DV[PVia].Fim) DO
					PVia := PVia + 1;

				{ Seleciona a menor chave }
				MenorVia := PVia;
				MenorKey := DV[PVia].R.Chave;
				PVia := PVia + 1;
				WHILE PVia <= Pv DO BEGIN
					IF NOT DV[PVia].Fim THEN
						IF DV[PVia].R.Chave < MenorKey THEN BEGIN
							MenorVia := PVia;
							MenorKey := DV[PVia].R.Chave;
							END;
					PVia := PVia + 1;
					END;

				{ Grava o tag da via com a menor chave
				  no arquivo de referência }
				Write(A_Out, DV[MenorVia].R.Tag);

				DV[MenorVia].N := DV[MenorVia].N - 1;

				IF DV[MenorVia].N = 0 THEN BEGIN
					DV[MenorVia].Fim := TRUE;
					V_Ativas         := V_Ativas - 1;
					END
				ELSE { Le mais um registro desta via }
					Read(DV[MenorVia].F, DV[MenorVia].R);
				END;

			{ Neste ponto apenas uma das vias
			  contém chaves cujos tags ainda não foram
			  transferidos para o arquivo de referência }

			{ Localiza a via com tags pendentes }
			PVia := 1;
			WHILE (PVia <= Pv) AND (DV[PVia].Fim) DO
				PVia := PVia + 1;

			WHILE(DV[PVia].N > 0) DO BEGIN
				Write(A_Out, DV[PVia].R.Tag);

				DV[PVia].N := DV[PVia].N - 1;

				IF DV[PVia].N > 0 THEN
					{ Le mais um registro desta via }
					Read(DV[PVia].F, DV[PVia].R);
				END;

			{ Fecha e remove todas as vias }
			FOR I := 1 TO Pv DO BEGIN
				Close(DV[I].F);
				Erase(DV[I].F);
				END;

			{ Fecha o arquivo de saída }
			Close(A_Out);

		END; { Fase }

	BEGIN
		{ Abre o arquivo de entrada }
		Assign(A_In,A);
		Reset(A_In);

		{ Ordena e distribui N chaves em P vias  }
		Pos := 0;
		FOR I := 1 TO Pv DO
			Distribui_F(I);

		{ Fecha o arquivo de entrada }
		Close(A_In);

		{ Realiza a operação de merge com as P vias
		  gerando o arquivo de referência }
		Fase;

	END; { Merge_P_Vias }

	PROCEDURE Simula_Dados;
	BEGIN

		{ Produz dados para ordenação }
		Assign(A_In,Entrada);
		ReWrite(A_In);

		Registro.Idade   := 18;
		Registro.Salario := 1000.0;

		{ Grave N registros sendo que o nome (chave) de cada
		  registro é composto por duas letras geradas em
		  ordem inversa }
		FOR I := 25 DOWNTO 0 DO BEGIN
			Registro.Nome[1] := Chr(I + Ord('A'));
			FOR J := 25 DOWNTO 0 DO BEGIN
				Registro.Nome[2] := Chr(J + Ord('a'));
				Write(A_In, Registro);
				Registro.Idade   := Registro.Idade + 1;
				Registro.Salario := Registro.Salario + 500.0;
				IF Registro.Idade > 69 THEN
					Registro.Idade := 18;
				IF Registro.Salario > 5000.00 THEN
					Registro.Salario := 1000.0;
				END;
			END;

		Close(A_In);

	END; { Simula_Dados }

	{ Verifica se as chaves estão em ordem }
	PROCEDURE Verifica_Ordenacao;
	VAR
		Tag : INTEGER;
	BEGIN
		{ Abre o arquivo de entrada }
		Assign(A_In,Entrada);
		Reset(A_In);

		{ Abre o arquivo de referência }
		Assign(A_Out,Saida);
		Reset(A_Out);

		{ Leia o tag do registro com a menor chave }
		Read(A_Out,Tag);

		{ Posicione o arquivo de entrada para ler
		  o registro associado com este tag }
		Seek(A_In,Tag - 1);

		{ Leia o registro completo do arquivo de entrada }
		Read(A_In,Registro);

		{ Verifica se a chave de um registro é
		  maior que a de sucessor para todos os registros }
		I := 2;
		WHILE I <= N DO BEGIN
			Read(A_Out,Tag);
			Seek(A_In,Tag - 1);
			Read(A_In,Registro2);
			IF Registro.Nome > Registro2.Nome THEN BEGIN
				WriteLn('O registro ', I, 'esta'' fora de ordem...');
				Writeln(Registro.Nome, ' ', Registro2.Nome, Tag - 1);
				Halt;
				END;
			Registro := Registro2;
			I := I + 1;
			END;

		Close(A_In);
		Close(A_Out);
		Erase(A_Out);

		WriteLn('A ordem esta'' correta!');

	END; { Verifica_Ordenacao }

BEGIN

	Simula_Dados;

	Merge_P_Vias(Entrada,Saida);

	Verifica_Ordenacao;

END. { MergePV }